### PR TITLE
Estimate functionality

### PR DESCRIPTION
**Pull request by grammeaway**

### CODE REVIEW

File: TaskHouseApi/Model/Estimate.cs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+            this.HourlyWage = HourlyWage;
+            this.Urgency = Urgency;
+
+            this.PriceEstimate = Calculate();
+        }
+
+        public decimal Calculate()
+        {
+            return TotalHours * HourlyWage * Complexity * Urgency;
+        }
```
Comment (kvartborg) [inline TaskHouseApi/Model/Estimate.cs]: PriceEstimate can get out of sync if calculate isn’t called before its used. I would recommend to remove PriceEstimate and use Calculate instead when the price is needed.

File: TaskHouseApi/Model/Task.cs [hunk 2]
```diff
@@ -29,12 +28,31 @@ public class Task : BaseModel
...
+
+            if (Estimates.Count > 0)
+            {
+                foreach (Estimate e in Estimates)
+                {
+                    SummedEstimate += e.PriceEstimate;
+                }
+
+                return SummedEstimate / Estimates.Count;
+            }
```
Comment (kvartborg) [inline TaskHouseApi/Model/Task.cs]: Turn if statement around, so the 0 case is handled before the calculation.

File: TaskHouseUnitTests/UnitTests/EstimateControllerUnitTest.cs [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+        public void EstimatesController_Get_ReturnAllElementsInRepo_WhenGivenNoParameters()
+        {
+            var result = controller.Get();
+            var resultObjectResult = result as ObjectResult;
+            var resultObject = resultObjectResult.Value as IEnumerable<Estimate>;
+
+            Assert.IsType<ObjectResult>(result);
+            Assert.Equal(4, resultObject.Count());
+        }
+
```
Comment (MadsDueKristensen) [inline TaskHouseUnitTests/UnitTests/EstimateControllerUnitTest.cs]: white

File: TaskHouseUnitTests/UnitTests/EstimateControllerUnitTest.cs [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+            //Arrange id for location object
+            int estimateId = 403;
+
+            //Act
+            var result = controller.Get(estimateId) as NotFoundResult;
+
+            //Assert
+            Assert.Equal(404, result.StatusCode);
+        }
+
```
Comment (MadsDueKristensen) [inline TaskHouseUnitTests/UnitTests/EstimateControllerUnitTest.cs]: white

File: TaskHouseUnitTests/UnitTests/EstimateControllerUnitTest.cs [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+        {
+            //Arrange
+            int Id = 100;
+
+            //Act
+            var result = controller.Delete(Id);
+
+            //Assert
+            Assert.IsType<NotFoundResult>(result);
+
```
Comment (MadsDueKristensen) [inline TaskHouseUnitTests/UnitTests/EstimateControllerUnitTest.cs]: white

File: TaskHouseUnitTests/UnitTests/TasksControllerUnitTests.cs [hunk 6]
```diff
@@ -238,5 +253,99 @@ public void TasksController_Delete_ReturnsNotFoundResult_WhenIdIsInvalid()
...
+                Complexity = 1,
+                HourlyWage = 110M,
+                Urgency = 1M,
+            };
+
+            int taskId = 500;
+
+            //Act
+            var result = controller.Create(taskId, e);
+
```
Comment (MadsDueKristensen) [inline TaskHouseUnitTests/UnitTests/TasksControllerUnitTests.cs]: white

File: TaskHouseUnitTests/UnitTests/TasksControllerUnitTests.cs [hunk 7]
```diff
@@ -238,5 +253,99 @@ public void TasksController_Delete_ReturnsNotFoundResult_WhenIdIsInvalid()
...
+
+            //Assert
+            Assert.IsType<BadRequestObjectResult>(result);
+
```
Comment (MadsDueKristensen) [inline TaskHouseUnitTests/UnitTests/TasksControllerUnitTests.cs]: white

File: TaskHouseUnitTests/UnitTests/TasksControllerUnitTests.cs [hunk 8]
```diff
@@ -238,5 +253,99 @@ public void TasksController_Delete_ReturnsNotFoundResult_WhenIdIsInvalid()
...
+            Estimate e = null;
+
+            int taskId = 1;
+
+            //Act
+            var result = controller.Create(taskId, e);
+
+            //Assert
+            Assert.IsType<BadRequestObjectResult>(result);
+
```
Comment (MadsDueKristensen) [inline TaskHouseUnitTests/UnitTests/TasksControllerUnitTests.cs]: white

File: TaskHouseUnitTests/UnitTests/TasksControllerUnitTests.cs [hunk 9]
```diff
@@ -238,5 +253,99 @@ public void TasksController_Delete_ReturnsNotFoundResult_WhenIdIsInvalid()
...
+        public void TasksController_Estimate_Create_AverageEstimateForTaskIszero_WhenGEstimateListIsEmpty()
+        {
+            //arrange
+            int taskId = 1;
+
+            //Act
+            var result = controller.Get(taskId);
+            var resultAsObject = controller.Get(taskId) as ObjectResult;
+            var resultObject = resultAsObject.Value as TaskHouseApi.Model.Task;
+
```
Comment (MadsDueKristensen) [inline TaskHouseUnitTests/UnitTests/TasksControllerUnitTests.cs]: white

File: TaskHouseApi/Model/Estimate.cs [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+        [Required]
+        public int WorkerId { get; set; }
+        [Required]
+        public int TaskId { get; set; }
+        public decimal PriceEstimate { get; set; }
+
+        public Estimate()
+        {
+
+        }
```
Comment (MadsDueKristensen) [inline TaskHouseApi/Model/Estimate.cs]: public Estimate() { }

File: TaskHouseApi/Persistence/Repositories/Interfaces/IEstimateRepository.cs [hunk 11]
```diff
@@ -0,0 +1,12 @@
+using System.Collections.Generic;
+using TaskHouseApi.Model;
+
+namespace TaskHouseApi.Persistence.Repositories.Interfaces
+{
+    public interface IEstimateRepository : IRepository<Estimate>
+    {
+        //Custom functionality to be added
+        IEnumerable<Estimate> RetrieveAllEstimatesForSpecificTaskId(int Id);
+
```
Comment (MadsDueKristensen) [inline TaskHouseApi/Persistence/Repositories/Interfaces/IEstimateRepository.cs]: white

**Pull request merged**